src/serializer/ResidualHeapSerializer.js

```diff
@@ -0,0 +1,1401 @@
+  >;
+  functions: Map<BabelNodeBlockStatement, Array<FunctionInstance>>;
+  functionInstances: Array<FunctionInstance>;
+  //value to intermediate references generated like $0, $1, $2,...
+  refs: Map<Value, BabelNodeIdentifier>;
+  collectValToRefCountOnly: boolean;
+  valToRefCount: void | Map<Value, number>;
+  prelude: Array<BabelNodeStatement>;
+  body: Array<BabelNodeStatement>;
+  mainBody: Array<BabelNodeStatement>;
+  realm: Realm;
+  declaredDerivedIds: Set<BabelNodeIdentifier>;
+  preludeGenerator: PreludeGenerator;
+  generator: Generator;
+  descriptors: Map<string, BabelNodeIdentifier>;
+  needsEmptyVar: boolean;
+  needsAuxiliaryConstructor: boolean;
+  valueNameGenerator: NameGenerator;
+  descriptorNameGenerator: NameGenerator;
+  factoryNameGenerator: NameGenerator;
+  logger: Logger;
+  modules: Modules;
+  requireReturns: Map<number | string, BabelNodeExpression>;
+  statistics: SerializerStatistics;
+  timingStats: TimingStatistics;
+  residualHeapInspector: ResidualHeapInspector;
+  residualValues: Map<Value, Set<Scope>>;
+  residualFunctionBindings: Map<FunctionValue, VisitedBindings>;
+  residualFunctionInfos: Map<BabelNodeBlockStatement, FunctionInfo>;
+  serializedValues: Set<Value>;
+  residualFunctions: ResidualFunctions;
+
+  _getBodyReference() {
+    return new BodyReference(this.body, this.body.length);
+  }
+
+  addProperties(
+    name: string,
+    obj: ObjectValue,
+    reasons: Array<string>,
+    alternateProperties: ?Map<string, PropertyBinding>,
+    objectPrototypeAlreadyEstablished: boolean = false
+  ) {
+    /*
+    for (let symbol of obj.symbols.keys()) {
+      // TODO #22: serialize symbols
+    }
+    */
+
+    // inject properties
+    for (let [key, propertyBinding] of alternateProperties || obj.properties) {
+      invariant(propertyBinding);
+      let desc = propertyBinding.descriptor;
+      if (desc === undefined) continue; //deleted
+      if (this.residualHeapInspector.canIgnoreProperty(obj, key)) continue;
+      invariant(desc !== undefined);
+      this._eagerOrDelay(this._getDescriptorValues(desc).concat(obj), () => {
+        invariant(desc !== undefined);
+        return this._emitProperty(name, obj, key, desc, reasons);
+      });
+    }
+
+    // inject properties with computed names
+    if (obj.unknownProperty !== undefined) {
+      let desc = obj.unknownProperty.descriptor;
+      if (desc !== undefined) {
+        let val = desc.value;
+        invariant(val instanceof AbstractValue);
+        this._eagerOrDelay(this._getNestedAbstractValues(val, [obj]), () => {
+          invariant(val instanceof AbstractValue);
+          this._emitPropertiesWithComputedNames(obj, val, reasons);
+        });
+      }
+    }
+
+    // prototype
+    this.addObjectPrototype(name, obj, reasons, objectPrototypeAlreadyEstablished);
+    if (obj instanceof FunctionValue) this.addConstructorPrototype(name, obj, reasons);
+
+    this.statistics.objects++;
```

Comment:
I find myself wondering if this over/under counts the objects that are serialized.